src/hammer-vlsi/test.py

```diff
@@ -1103,25 +1103,147 @@ def change3(x: hammer_vlsi.HammerTool) -> bool:
 
     def test_resume_pause_hooks_with_custom_steps(self) -> None:
         """Test that resume/pause hooks work with custom steps."""
+        def step5(x: hammer_vlsi.HammerTool) -> bool:
+            with open(os.path.join(c.temp_dir, "step5.txt"), "w") as f:
+                f.write("HelloWorld")
+            return True
+        def step6(x: hammer_vlsi.HammerTool) -> bool:
+            with open(os.path.join(c.temp_dir, "step6.txt"), "w") as f:
+                f.write("ByeByeWorld")
+            return True
+
         with self.create_context() as c:
-            def step5(x: hammer_vlsi.HammerTool) -> bool:
-                with open(os.path.join(c.temp_dir, "step5.txt"), "w") as f:
-                    f.write("HelloWorld")
-                return True
+            # Test from_step => to_step
+            c.driver.set_post_custom_syn_tool_hooks(hammer_vlsi.HammerTool.make_start_stop_hooks("step5", None, "step5", None))
+            success, syn_output = c.driver.run_synthesis(hook_actions=[
+                hammer_vlsi.HammerTool.make_post_insertion_hook("step4", step5),
+                hammer_vlsi.HammerTool.make_post_insertion_hook("step5", step6)
+            ])
+            self.assertTrue(success)
 
-            c.driver.set_post_custom_syn_tool_hooks(hammer_vlsi.HammerTool.make_from_to_hooks("step5", "step5"))
+            for i in range(1, 7):
+                file = os.path.join(c.temp_dir, "step{}.txt".format(i))
+                if i == 5:
+                    self.assertEqual(self.read(file), "HelloWorld")
+                else:
+                    self.assertFalse(os.path.exists(file))
+
+        with self.create_context() as c:
+            # Test after_step => until_step
+            c.driver.set_post_custom_syn_tool_hooks(hammer_vlsi.HammerTool.make_start_stop_hooks(None, "step4", None, "step6"))
             success, syn_output = c.driver.run_synthesis(hook_actions=[
-                hammer_vlsi.HammerTool.make_post_insertion_hook("step4", step5)
+                hammer_vlsi.HammerTool.make_post_insertion_hook("step4", step5),
+                hammer_vlsi.HammerTool.make_post_insertion_hook("step5", step6)
             ])
             self.assertTrue(success)
 
-            for i in range(1, 6):
+            for i in range(1, 7):
                 file = os.path.join(c.temp_dir, "step{}.txt".format(i))
                 if i == 5:
                     self.assertEqual(self.read(file), "HelloWorld")
                 else:
                     self.assertFalse(os.path.exists(file))
 
+    def test_persistent_hooks(self) -> None:
+        """Test that persistent hooks work."""
+        def persist(x: hammer_vlsi.HammerTool) -> bool:
```

Comment:
ditto

Reply:
ditto comment above, it would duplicate code.